test/TestNestedJinjaRule.py

```diff
@@ -0,0 +1,184 @@
+  set_fact:
+    var_two: "2*(1+(3-1)) is {{ 2 * {{ 1 + {{ 3 - 1 }} }} }}"
+''')
+
+FAIL_TASK_2LN_M = PlayFile('playbook.yml', u'''
+- name: two-level multiline nesting
+  set_fact:
+    var_two_ml: >
+      2*(1+(3-1)) is {{ 2 *
+      {{ 1 +
+      {{ 3 - 1 }}
+      }} }}
+''')
+
+FAIL_TASK_W_5LN = PlayFile('playbook.yml', u'''
+- name: five-level wild nesting
+  set_fact:
+    var_three_wld: "{{ {{ {{ {{ {{ 234 }} }} }} }} }}"
+''')
+
+FAIL_TASK_W_5LN_M = PlayFile('playbook.yml', u'''
+- name: five-level wild multiline nesting
+  set_fact:
+    var_three_wld_ml: >
+      {{
+      {{
+      {{
+      {{
+      {{ 234 }}
+      }}
+      }}
+      }}
+      }}
+''')
+
+SUCCESS_TASK_P = PlayFile('playbook.yml', u'''
+- name: non-nested example
+  set_fact:
+    var_one: "number for 'one' is {{ 2 * 1 }}"
+''')
+
+SUCCESS_TASK_P_M = PlayFile('playbook.yml', u'''
+- name: multiline non-nested example
+  set_fact:
+    var_one_ml: >
+      number for 'one' is {{
+      2 * 1 }}
+''')
+
+SUCCESS_TASK_2P = PlayFile('playbook.yml', u'''
+- name: nesting far from each other
+  set_fact:
+    var_two: "number for 'two' is {{ 2 * 1 }} and number for 'three' is {{ 4 - 1 }}"
+''')
+
+SUCCESS_TASK_2P_M = PlayFile('playbook.yml', u'''
+- name: multiline nesting far from each other
+  set_fact:
+    var_two_ml: >
+      number for 'two' is {{ 2 * 1
+      }} and number for 'three' is {{
+      4 - 1 }}
+''')
+
+SUCCESS_TASK_C_2P = PlayFile('playbook.yml', u'''
+- name: nesting close to each other
+  set_fact:
+    var_three: "number for 'ten' is {{ 2 - 1 }}{{ 3 - 3 }}"
+''')
+
+SUCCESS_TASK_C_2P_M = PlayFile('playbook.yml', u'''
+- name: multiline nesting close to each other
+  set_fact:
+    var_three_ml: >
+      number for 'ten' is {{
+      2 - 1
+      }}{{ 3 - 3 }}
+''')
+
+
+@pytest.fixture
+def runner(tmp_path, default_rules_collection):
+    return Runner(
+        default_rules_collection,
+        str(tmp_path / 'playbook.yml'),
+        [], [], [],
+    )
+
+
+@pytest.fixture
+def _playbook_file(tmp_path, request):
+    if request.param is None:
+        return
+    for play_file in request.param:
+        p = tmp_path / play_file.name
+        p.write_text(play_file.content)
+
+
+@pytest.mark.parametrize(
+    '_playbook_file',
+    (
+        pytest.param([FAIL_TASK_1LN], id='file includes one-level nesting'),
+        pytest.param([FAIL_TASK_1LN_M], id='file includes one-level multiline nesting'),
+        pytest.param([FAIL_TASK_2LN], id='file includes two-level nesting'),
+        pytest.param([FAIL_TASK_2LN_M], id='file includes two-level multiline nesting'),
+        pytest.param([FAIL_TASK_W_5LN], id='file includes five-level wild nesting'),
+        pytest.param([FAIL_TASK_W_5LN_M], id='file includes five-level wild multiline nesting'),
+    ),
+    indirect=['_playbook_file'],
+)
+@pytest.mark.usefixtures('_playbook_file')
+def test_including_wrong_nested_jinja(runner):
+    results = runner.run()
+    # FIXME: results == []
```

Comment:
why?

Reply:
**The runner does not return any value in the tests.**

It always returns only an empty list, whether the input is correct or incorrect. When I check if the mocked file includes the given rule, it does exist as it is supposed to be. The new rule is also included in that runner as it is expected.

The rule always returns a boolean data type but it cannot be cached in the test for now.

Whenever I launch ansible-lint as `ansible-lint -r <DIR> <FILE>` while the `<DIR>` includes that new rule, it works perfectly based on the correct/incorrect `<FILE>` input.

Should I change the method of runner launching?

Reply:
Not sure what you mean. Are you saying that it doesn't work with the default rules dir?

Reply:
The created runner object includes all the rules (including the 'nested jinja rule') as well as the mocked file, which is properly created.
```
(Pdb) runner.checked_files
{'../../../../tmp/pytest-of-adrian/pytest-0/test_including_wrong_nested_ji0/playbook.yml'}
(Pdb) import os; os.system("cat ../../../../tmp/pytest-of-adrian/pytest-0/test_including_wrong_nested_ji0/playbook.yml")

- name: one-level nesting
  set_fact:
    var_one: "2*(1+2) is {{ 2 * {{ 1 + 2 }} }}"
0
(Pdb) runner.rules
101: ...
...
206: ...
  ...
207: Nested jinja pattern
  There should not be any nested jinja pattern. Example (bad): ``{{ list_one + {{ list_two | max }} }}``, example (good): ``{{ list_one + max(list_two) }}``
301: ...
  ...
...
901: ...
  ...
```

I use `runner.run()` accordingly to [`Runner` definition](https://github.com/ansible/ansible-lint/blob/master/lib/ansiblelint/runner.py#L56)
```
class Runner(object):
    ...
    def run(self) -> List:
    ...
```

The runner should return a nonempty list in case of "bad" input. Do I launch the runner incorrectly in the tests that I always have an empty list?

Reply:
Looks about right. Have you tried sticking debugger into your rule's `matchtask()`?

Reply:
Have you tried converting tasks lists into actual playbooks?
(I mean add one more level of nesting and put tasks under `tasks:`)

Reply:
The `matchtask` behavior has changed, I had to fix it again. Now, the runner in the tests returns a list of errors as expected. This thread has been resolved.